x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -277,126 +398,126 @@ private static XContentBuilder getDynamicTemplates(XContentBuilder builder) thro
-                Index rollupIndex = rollupIndexMetadata.getIndex();
-                IndexAbstraction originalIndex = currentState.getMetadata().getIndicesLookup().get(originalIndexName);
-
-                Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
-                if (originalIndex.getParentDataStream() != null) {
-                    // If rolling up a backing index of a data stream, add rolled up index to backing data stream
-                    DataStream originalDataStream = originalIndex.getParentDataStream().getDataStream();
-                    List<Index> backingIndices = new ArrayList<>(originalDataStream.getIndices().size() + 1);
-                    // Adding rollup indices to the beginning of the list will prevent rollup indices from ever being
-                    // considered a write index
-                    backingIndices.add(rollupIndex);
-                    backingIndices.addAll(originalDataStream.getIndices());
-                    DataStream dataStream = new DataStream(
-                        originalDataStream.getName(),
-                        backingIndices,
-                        originalDataStream.getGeneration(),
-                        originalDataStream.getMetadata(),
-                        originalDataStream.isHidden(),
-                        originalDataStream.isReplicated(),
-                        originalDataStream.isSystem(),
-                        originalDataStream.isAllowCustomRouting(),
-                        originalDataStream.getIndexMode()
-                    );
-                    metadataBuilder.put(dataStream);
+            public void onFailure(Exception e) {
+                listener.onFailure(new ElasticsearchException("Failed to finalize the rollup state for index [" + rollupIndex + "]", e));
+            }
+        });
+    }
+
+    private void refreshIndex(String index, ActionListener<AcknowledgedResponse> listener) {
+        client.admin()
+            .indices()
+            .refresh(
+                new RefreshRequest(index),
+                ActionListener.wrap(
+                    refreshResponse -> listener.onResponse(AcknowledgedResponse.TRUE),
+                    e -> listener.onFailure(new ElasticsearchException("Failed to refresh index [" + index + "]", e))
+                )
+            );
+    }
+
+    private void deleteSourceIndex(final String sourceIndex, final String rollupIndex, ActionListener<AcknowledgedResponse> listener) {
+        client.admin().indices().delete(new DeleteIndexRequest(sourceIndex), new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                if (acknowledgedResponse.isAcknowledged()) {
+                    // Source index was deleted successfully.
+                    // 8. Finalize the rollup index by setting its state to SUCCESS
+                    finalizeRollupIndex(rollupIndex, listener);
```

Comment:
Is it possible to submit the delete action and for the source index to be deleted and then something to go wrong and you don't get the response here so the rollup index isn't "finalized"? I think it'd be nice to make the replacement in the data stream, the delete, and the finalize atomic in one action.

Reply:
I modified the code so that the following operations happen within a single state update request, so they are all performed atomically.

 - If source index belongs to a data stream, add the rollup index and remove the source index
 - Mark rollup index as completed successfully
 - Delete the source index


See [ee42ef5](https://github.com/elastic/elasticsearch/pull/85708/commits/ee42ef5aacc5328ceef64b4f052f2e11a0537f74)